### ISSUE TITLE
Must be missing something

### ISSUE THREAD

**Issue by FredericLatour** [nestjsx/nestjs-config#95]
**Issue type:**

- [x] question
- [ ] bug report
- [ ] feature request
- [ ] documentation issue

**nestjs-config version**

**@nestjs/common+core or other package versions**
 - @nestjs/common: 
 - @nestjs/core: 

#### Excepted behavior 
I must be missing something but using your approach you are mostly losing the benefit of types in typescript.
Wouldn't be possible to provide config tools without a get function in the middle and by accessing typed config properties directly?

#### Replication/Example

```ts
```

**Comment by bashleigh** [comment]
Yea it's possible, you can use your own class and implement an interface as well as extend the Config class to use the get method. 

I see your point about 'losing the benefit of types' because the ENVs themselves are usually string | number. The main reason I added the ability to use classes was for interfaces and DI into other modules. The second was moving onto validation at runtime and the third (suggested by @shekohex ) to create type.env files that 'auto' convert the given ENVs into a type where necessary.

I released version 2.0.0-beta to get feedback from users about the new 'features' and features I had removed/replaced. Adding types was a bit of an extra benefit that I envisioned like so:

```ts
//src/actions/action.module.interface.ts
export interface IAction {
  name: string;
  amount: number;
}
export ActionModuleOptions {
  actions: IAction[];
}

//src/actions/action.module.ts
@Module({})
export ActionModule {
  forRootAsync(options: {
    inject?: Provider[],
    useFactory: (...args: any) => ActionModuleOptions,
   }): DynamicModule {
     //...do something with options.actions
  }
}

//src/config/actions.ts
export class ActionOptions implements ActionModuleOptions {
  public actions: [];
  constructor() {
    const actions = process.env.ACTIONS.split(',');
    actions.forEach(action => action.amount = parseInt(action.amount ? action.amount : 0));
  }
}

//src/main.ts

@Module({
  imports: [
    ConfigModule.forRootAsync(path.resolve(__dirname, 'config', '**/!(*.d).{ts,js}'),
    ActionModule.forRootAsync({
      useFactory: (config: ActionModuleOptions) => config, //no config.get as get() doesn't exist on ActionModuleOptions or ActionOptions
      inject: [ActionModuleOptions],
    }),
  ],
})
export class AppModule {}
```

I see your point about injecting typed properties directly. In 2.0.0-beta I have removed the decorators to 'inject' properties into methods. I was considering developing a replacement however I've struggled to come up with a nice method of doing so. 

What do you think about:

1. The above example
2. Runtime type validation of provided `process.env.{requested}`
3. type.env file + auto conversion 
4. Injecting config parameters/properties

**Comment by FredericLatour** [comment]
@bashleigh Thanks for getting back. To be honest I'm not completely sure how I would like a config system to work. This something I'm thinking of for quite some time without being completely sure of what the best approach could be.

Currently, I'm using [node-config](https://github.com/lorenwest/node-config) in a specific way. I rely on the module to get my config object (based on json5 files) based on my current env (dev, prod, etc ...).   
However instead of using their `config.get` method, I'm using directly `config as IMyConfigInterface` so that I can have full intellisense over my entire config.  
This is a bit hacky I have to admit because I have to maintain a proper interface and there is no validation of my json5 file over my interface. That said, I value having intellisense for my config.

Moreover, my approach is far from perfect on a functional and secure standpoint in the sense that there should be 2 kind of configuration data:

- Internal configuration  data that we don't necessarily want to change externally and that we would like to commit in the source.
- External configuration data that is supposed to be set through environment (the exact mechanism will depends on the platform : kubernetes, docker-compose, etc ...) and should not be committed:. Login, passwords, keys, etc ....

So basically, in an ideal world, I have a combination of config through files (json? ts?) and environment with some validation mechanism and still having intellisense.


Now some questions may arise on a conceptual standpoint:

- Should a module (or service or whatever you call it) config information be private to this module?
- Is cross cutting config data necessarily supposed to be configured through env and not be commited or should we have both ? 

For information, I'm not using nestjs yet. I'm just testing it.  I came across your module while searching if nestjs had an opinionated approach to solve the config problem.

**Comment by bashleigh** [comment]
When @fenos and I started this repo, we both agreed that what we was both trying to achieve was a config system similar to Laravel the PHP framework which uses several files within a folder called config. The name of your file is then used for getting a key within the file like `config('file-name.key');`.

This package have evolved since then and many developers have left feedback and features requires as well as contributing to make this config package suit their needs or requirements. 

Me personally, I've used, use this package within kubernetes containers, docker containers and on servers!  I use environment variables on the container/server which I then use dotenv to fetch into my process and this config package to restructure, and organise my configurations accordingly. Quite often than not now, I'm using this package more and more for configuring modules rather than using the configService within those module's themselves. 
So my goal with the V2 release was to make the config package more 'nestjs' like and apply more SOILD principles to the package by added the `forRootAsync` method and enabling the module to 'require' config files, make a configurable class in the file and use Dependancy Injection with that class. 
The drawback however is that nestjs is asynchronous meaning the class I'm requiring within the config package doesn't exist until AFTER the config module has successfully finished being instanced. So we need to 'await' the config class to be created before we can successfully create our module. Which is correct for other module which await other modules to be instanced which configuring them in 'imports'. 

I would believe that a module's config should be private to that module. I don't want to use the TypeOrmModule's config elsewhere in my application. 

In regards to committing configs, I usually make a `.env.dist` or `.env.example` file for development. So let's say my friend wants to start working on my application; to save himself time he can clone the repo and copy the [`.env.dist`](https://github.com/bashleigh/nestjs-blog/blob/master/.env.dist) file. Sometimes I include a test and cp for the .env file. .env.dist file usually just contain the env names and sometimes default values for development environments. 
I wouldn't commit private env's like your database password, application secret (for password encrypting etc). That stuff stays private. In kubernetes they're usually 'saved' as secrets and configured into the container's environment variables. 

The nestjs-config package should cover both your implementation and interfaces in V2. If you make a class within your config you can import it and use that as your interface/intellisense. 

Hope this helps! If you have to have a specific setup or an example I can help you with me know and I'll do my best to help out!

**Comment by metauro** [comment]
@FredericLatour @bashleigh 
Hi, what do you think for this mode? 
https://github.com/metauro/nest-configuration#readme

**Comment by bashleigh** [comment]
@metauro I like it! It's pretty simple :) How would you define secret keys though?

**Comment by metauro** [comment]
@bashleigh a simple way is just create a config template file, and then git ignore all config or private config file but don't ignore template file. 

e.g.
```typescript
// database-config.service.ts

@Configuration
export class DatabaseConfigService { 
  @Config({
    base: 'root'
  })
  user: string;

  @Config({
    base: 'root'
  })
  password: string;
}

// database-config.service.template.ts
@Configuration
export class DatabaseConfigService {
  @Config({
    base: 'template value'
  })
  user: string;
  
  @Config({
    base: 'template value'
  })
  password: string;
}

// .gitignore
src/*-config.service.ts
```

for another way, we could encrypt config file, but I don't know how to do this now.
do you have any ideas?
